src/invariant/OrderBookIsNotCrossed.cpp

```diff
@@ -13,81 +13,31 @@
 namespace stellar
 {
 
-std::vector<std::pair<Asset, Asset>>
-extractAssetPairs(Operation const& op, LedgerTxnDelta const& ltxd)
+static std::set<std::pair<Asset, Asset>>
+extractAssetPairs(LedgerTxnDelta const& ltxd)
 {
-    switch (op.body.type())
-    {
-    case MANAGE_BUY_OFFER:
-    {
-        auto const& offer = op.body.manageBuyOfferOp();
-        return {std::make_pair(offer.selling, offer.buying)};
-    }
-    case MANAGE_SELL_OFFER:
-    {
-        auto const& offer = op.body.manageSellOfferOp();
-        return {std::make_pair(offer.selling, offer.buying)};
-    }
-    case CREATE_PASSIVE_SELL_OFFER:
-    {
-        auto const& offer = op.body.createPassiveSellOfferOp();
-        return {std::make_pair(offer.selling, offer.buying)};
-    }
-    case PATH_PAYMENT:
-    {
-        auto const& pp = op.body.pathPaymentOp();
-
-        // if no path, only have a pair between send and dest
-        if (pp.path.size() == 0)
-        {
-            return {std::make_pair(pp.sendAsset, pp.destAsset)};
-        }
-
-        // For send, dest, {A, B} we get: {send, A}, {A, B}, {B, dest}
-        std::vector<std::pair<Asset, Asset>> assets;
-
-        // beginning: send -> A
-        assets.emplace_back(pp.sendAsset, pp.path.front());
-        for (int i = 1; i < pp.path.size(); ++i)
-        {
-            // middle: A -> B
-            assets.emplace_back(pp.path[i - 1], pp.path[i]);
-        }
-        // end: B -> dest
-        assets.emplace_back(pp.path.back(), pp.destAsset);
 
-        return assets;
-    }
-    case ALLOW_TRUST:
+    std::set<std::pair<Asset, Asset>> assets;
+    for (auto const& entry : ltxd.entry)
     {
-        auto const& at = op.body.allowTrustOp();
-
-        // if revoke auth, all offers for that user against that asset are
-        // deleted
-        if (!at.authorize)
+        if (entry.first.type() == OFFER)
         {
-            std::vector<std::pair<Asset, Asset>> assets;
-            // since we only get one side of the asset pair from the operation,
-            // we derive the rest of the information from the
-            // LedgerTxnDelta entries
-            for (auto const& entry : ltxd.entry)
+            auto const& offer = entry.second.previous
```

Comment:
Looking closely at this, I'm not sure that this makes sense as implemented. An important piece of context here: it _is_ possible to modify the `buying` or `selling` assets on an offer. In those cases, `previous` and `current` won't have the same asset pairs.

But in any case, I'm wondering why we should ever care about the `previous` state? If an order book wasn't crossed and an order was deleted, then it still isn't crossed. If an order book wasn't crossed and an order was modified such that it has the same asset pair, then it will be gathered from the current state. If an order book wasn't crossed and an order was modified such that it doesn't have the same asset pair, then it still isn't crossed. In any of these cases, the `previous` asset pair isn't useful. Am I missing something here?

Reply:
Thanks, I did not realize you can change the asset, I thought you could only change the price and amount. Good catch.

Very good point -- if a `set A` of orders on an order book is not crossed, there is **no** `subset B` such that it may become crossed. The spread between the lowest ask and the highest bid for the subset `B` is `>=` the spread before the deletion of any order `A`.

Reply:
**TLDR** they can only be in a less or equally crossed state so if the state was not crossed before it is not crossed now.